src/fromager/commands/bootstrap.py

```diff
@@ -77,19 +88,47 @@ def bootstrap(
         )
     logger.info("bootstrapping %r variant of %s", wkctx.variant, to_build)
 
+    prev_graph = None
+    if prev_graph_file:
+        prev_graph = dependency_graph.DependencyGraph.from_file(prev_graph_file)
+
     pre_built = wkctx.settings.list_pre_built()
     if pre_built:
         logger.info("treating %s as pre-built wheels", sorted(pre_built))
 
     server.start_wheel_server(wkctx)
 
+    # we need to resolve all the top level dependencies before we start bootstrapping.
+    # this is to ensure that if we are using an older bootstrap to resolve packages
+    # we are able to upgrade a package anywhere in the dependency tree if it is mentioned
+    # in the toplevel without having to fall back to history
+    for req in to_build:
+        pbi = wkctx.package_build_info(req)
+        if pbi.pre_built:
+            servers = wheels.get_wheel_server_urls(wkctx, req)
+            source_url, version = wheels.resolve_prebuilt_wheel(wkctx, req, servers)
+        else:
+            source_url, version = sources.resolve_source(
+                wkctx, req, resolver.PYPI_SERVER_URL
+            )
```

Comment:
What other cases do we deal with in other places? Does this single `if` cover everything now? Your comments last week about reusing the resolver have me wondering about how much benefit we would get from sharing this logic in other places by creating a function?

Reply:
the only other place we do this `if` based resolution is in sdist.py: https://github.com/python-wheel-build/fromager/pull/399/files/810230839452154a7b3b284a3a7d8a6ab57c51b0#diff-3f261cb4d56f2826fc58f44967dd7d1b2a92b5e4e9da0133b832d0e9027ff011R57

Reply:
Ok,it felt like it might have been a pattern but it sounds like not.